src/conv-core/convcore.C

```diff
@@ -3031,12 +3031,33 @@ void CmiGroupInit(void)
 
 #if CMK_MULTICAST_LIST_USE_COMMON_CODE
 
-void CmiSyncListSendFn(int npes, const int *pes, int len, char *msg)
+void CmiSyncListSendFn(int npes, const int* pes, int len, char* msg)
 {
-  int i;
-  for(i=0;i<npes;i++) {
-    CmiSyncSend(pes[i], len, msg);
+  // When in SMP mode, each send needs its own message, there is a race between unpacking
+  // for local PEs and there is a race between setting the rank in the Converse header and
+  // the actual comm thread send for remote PEs
+#if CMK_SMP
+  for (int i = 0; i < npes - 1; i++)
+  {
+    const char* msgCopy = CmiCopyMsg(msg, len);
+    CmiSyncSendAndFree(pes[i], len, msgCopy);
   }
+  // No need to copy and free for the last send, so use the original message
+  if (npes > 0)
+    CmiSyncSend(pes[npes - 1], len, msg);
+#else
+  for (int i = 0; i < npes; i++)
+  {
+    // Copy if this is a self send to avoid unpack/send race
+    if (pes[i] == CmiMyPe() && npes > 1)
```

Comment:
Why would this cause a race? Wouldn't the unpack happen after I've finished doing all of the sends?

Reply:
Not necessarily, at least on some layers, even for non-SMP mode, the messaging layer still works by adding messages to a queue which it then processes between running EPs, but it checks its local queue before sending messages for external PEs (`src/arch/util/machine-common-core:1812-1816`):
```
    msg = CMIQueuePop(cs->recv); // checks local queue for msg
#endif
#if (!CMK_SMP || CMK_SMP_NO_COMMTHD) && !CMK_MULTICORE
    if (!msg) {
       AdvanceCommunication(0); // sends queued messages to remote PEs
```

When this happens, it is unpacked locally and then the same unpacked message is used for sends to remote PEs.

Reply:
Gotcha